### DISCUSSION TITLE
Folding files and persisting folds

### DISCUSSION THREAD

**Discussion by sdondley** [LazyVim/LazyVim#1358] [Recipes]
Here is a recipe for setting up folding with LazyVim.

1. Turn folding on:

```
vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"
vim.opt.foldmethod = "expr"
vim.opt.foldnestmax = 1
```

I set fondnextmax to 1 so only the top level functions are folded. This works for me but change as you see fit.

2.  To allow folds to be remembered across sessions add the following lua code to your config:

local autocmd = vim.api.nvim_create_autocmd
local augroup = vim.api.nvim_create_augroup
local save_fold = augroup("Persistent Folds", { clear = true })

```
autocmd("BufWinLeave", {
  pattern = "*.*",
  callback = function()
    vim.cmd.mkview()
  end,
  group = save_fold,
})
autocmd("BufWinEnter", {
  pattern = "*.*",
  callback = function()
    vim.cmd.loadview({ mods = { emsg_silent = true } })
  end,
  group = save_fold,
})
```

3. Enjoy.

**Comment by cStralpt** [comment]
it would be cool if we have unfolded behavior by default instead of folding the whole document, and ```za``` to toggle unfold/fold for every scope of the object/loop/function/html tags(detected automatically) just like lunarvim does

**Reply by OneOfOne** [reply]
@cStralpt `vim.opt.foldlevel = 99`
